week_1/Array/sort_colors/sort_colors.js

```diff
@@ -0,0 +1,37 @@
+// Given an array nums with n objects colored red, white, or blue, sort them in-place so that objects of the same color are adjacent, with the colors in the order red, white, and blue.
+
+// We will use the integers 0, 1, and 2 to represent the color red, white, and blue, respectively.
+// You may not use the sort method.
+
+// Warning: This algorithm mutates the input array.
+/**
+ * @param {number[]} nums
+ * @return {void} Do not return anything, modify nums in-place instead.
+ */
+const sortColors = (nums) => {
+	let low = 0;
+	let current = 0;
+	let high = nums.length - 1;
+
+	while (current <= high) {
+		// This swap ensures that every value whose index < low is equal to 0
+		if (nums[current] == 0) {
+			swap(low, current);
+			low++;
+			current++;
+		}
+		// This swap ensures that every value whose index > high is equal to 2
+		else if (nums[current] == 2) {
+			swap(current, high);
+			high--;
+		} else {
+			current++;
+		}
+	}
+
+	function swap(a, b) {
```

Comment:
A few things to consider here:
- You should always strive to have `pure functions`. It helps with the testability and the portability of your code. `swap` here is not considered pure, because it is referencing `nums` directly. To make this pure, you'd want this function to accept `nums` as an argument. Once `swap` expects `nums` to be supplied as an argument, you could begin to use `swap` anywhere you needed to swap some indexes around. You can also write tests just for `swap`, helping your codebase to have stronger test coverage.
- The only advantage I see of having `swap` defined within the function, as it is currently, is you've already defined your bounds. You know `a` and `b` are always going to be real indexes on your array (you presumably won't ever set low to -1, for instance). While one _could_ see this an advantage, I'd argue this is bad design. Code like this is akin to a tower of cards: `swap` assumes these indexes are within valid boundaries of `nums`. As soon as you refactor this code a bit, that might not be true. If this were a part of a larger code base, you'd begin to see breaking behavior that you might not have seen coming; like taking one card from one side of the tower, the rest of them begin to fall. To get around this, it's best practice when indexing an array to ensure the indexes you're referencing are valid. You'd want to ensure `a >= 0 && a < nums.length`. You'd do the same for `b.` You'd also want to ensure `a !== b`, as swapping those values around when you don't need to be is a waste of resources. And lastly, you'd want to make sure `a` and `b` are actually numbers. Since this is JavaScript, there is nothing stopping you from giving you this function strings instead of numbers. Another reason why its so easy to build towers of cards in JavaScript 😅 . At the very least, a JSDoc comment would discourage that from happening.

Reply:
Heard. I removed the swap function in favor of the ES6 destructuring method.